### ISSUE TITLE
⚠️ WebAssembly validation

### ISSUE THREAD

**Issue by vshymanskyy** [wasm3/wasm3#344]
Currently, Wasm3 assumes that input wasm files are validated. It performs many checks, but the Wasm3 validation process is incomplete.

As a result, we get quite [a lot of fuzzer errors](https://github.com/wasm3/wasm3/issues?q=is%3Aissue+is%3Aclosed+label%3Awasm-validation).

Validation should be performed according to the specification: https://github.com/WebAssembly/spec/archive/refs/tags/opam-1.1.1.zip

**Comment by vshymanskyy** [comment]
The validation has been implemented in a separate branch at the moment:
https://github.com/wasm3/wasm3/tree/validation

Feel free to test it and provide feedback.

> [!IMPORTANT]
> Please reopen any bugs that could have been closed by mistake.

**Comment by yulinlina** [comment]
I took a small documentation pass instead of touching the parser, since full spec validation is a much larger change. I added VALIDATION.md explaining that Wasm3 is not a complete validator and recommending external validation before running untrusted modules, and I'd also add a short pointer to it from README.md. This doesn't change runtime behavior, but it should reduce foot-guns while a proper validation implementation is being worked on.

**Comment by yulinlina** [comment]
That’s good news. If the goal is to close the validation bugs, I’d treat “AI thinks so” as a starting point, not an oracle. The dangerous failures are modules that pass parsing but violate the spec, so we need negative conformance evidence before merging.

A useful merge checklist would be:

- Run the official `wast2json` spec tests against the `validation` branch and require every invalid module to be rejected with a clean parse/validation error, not a runtime fault.
- Add regression binaries for previously reported fuzzer crashes: truncated sections, bad LEB128, out-of-range type/function/local/global indices, invalid control-flow depth, mismatched block signatures, bad immediates, invalid table/memory limits, malformed import/export entries, non-constant global initializers, and malformed element/data segments.
- Ensure validation happens before compilation/execution and returns errors distinguishable from I/O, OOM, or unsupported-feature errors.
- Decide whether validation is unconditional. For an interpreter that may receive untrusted modules, strict validation should probably be the default. If there is a performance-oriented “trusted input” path, it should be explicit rather than implicit.

I can help by running the branch against spec tests and old fuzzer artifacts, then collecting a list of still-accepted invalid modules. If there are failures, I can reduce them to minimal `.wasm` test cases for regression coverage. The docs note can also be adjusted to point at the validation branch once it’s the recommended path.

**Comment by vshymanskyy** [comment]
The goal is to merge validation into main once it's ready.
There is a flag to enable validations (Enabled by default).
That flag would only affect complicated validations, the "cheap" validations are always performed.

No, I don't think it will be possible to retrospectively validate everything. It should be enough to stick to the spec tests, the Cluster Fuzz will validate everything else once it's merged into the main. Whatever leftovers would have to be fixed later.

Otherwise this will never land

**Comment by vshymanskyy** [comment]
:tada: Most of the wasm-spec validations now PASS on all targets: https://github.com/wasm3/wasm3/actions?query=branch%3Avalidation
@soundandform FYI

**Comment by Robbepop** [comment]
Great to see that Wasm3 will be receiving proper Wasm validation support! 🚀

**Comment by yulinlina** [comment]
Agreed — this is the important part. The validation branch passing spec tests is a strong signal, but the merge criterion should be that malformed modules fail early with a validation error rather than later as undefined interpreter behavior.

After it lands, I’ll open a PR with a small regression corpus for the old fuzzer crash classes — bad LEB128, out-of-range indices, invalid control-flow depth, mismatched block signatures, malformed table/memory limits, non-constant globals, and bad element/data segments — plus a short documentation note that untrusted modules should still be validated externally until validation coverage is complete.

ClusterFuzz will pick up the long tail, but checked-in negative tests will make regressions bisectable.
